ath79: add support for Bluesocket BSAP1880 board

danielg4 opened the pull request:
This is in response to an appeal for device support from @bgonyer on IRC.
Forum thread: https://forum.openwrt.org/t/support-for-adtran-bluesocket-bsap-1800v2/10660

Both commits may potentially be needed to support this device. I began this effort before the #1237 merge, which introduces basic prerequisites for ath79 support for this device also, and ath79 was reported to be not quite ready, so I'm adding it as ar71xx. Since ar71xx is likely to go away within one release cycle, I figure the Kconfig ugliness, hackish Ethernet fix, and use of legacy.mk can be overlooked in favor of something that works today and can subsequently be converted to ath79 more cleanly. Corrections welcome.

danielg4 committed:
mtd: add CRC signature to RedBoot partition map

The code for calculating the CRC32 signatures for RedBoot FIS partitions
was already included, but for unknown reasons, it was never invoked. Some
bootloaders enforce checking these for loaded kernels, so they should be
written. This patch does so.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>
Tested-by: Brian Gonyer <bgonyer@gmail.com>

danielg4 committed:
ath79: add support for Bluesocket BSAP1880 board

This board was used in dual-band 802.11n enterprise access points, models
BSAP-1800v2 and BSAP-1840, introduced in 2010 by Bluesocket, which was
acquired by Adtran in 2011, who has now EOL'ed them. They differed only in
that the BSAP-1840's antennae were detachable, while the BSAP-1800v2's were
inside the case. They have an external RJ-45 console port, which works with
standard Cisco 72-3383-01 console cables.

Specification:
- System-On-Chip: AR7161
- CPU/Speed: 600 MHz
- Flash-Chip: Macronix MX25L12845E
- Flash size: 16 MiB
- RAM: 64 MiB
- Wireless No1: Lite-On WN2601A card: AR9160/AR9103 2.4GHz 802.11bgn
- Wireless No2: Lite-On WN2502A card: AR9160/AR9106 5GHz 802.11an
- PHY: Vitesse VSC8601, Rev. B

Installation:
1. Connect to the serial console using a terminal that supports YMODEM at
   115200 bps, 8 data bits, no parity, 1 stop bit
2. Interrupt the bootloader using its password, which is: r00t
3. Issue the "fis init" command, confirming if prompted
4. Look at the length of the openwrt-ath79-generic-*-squashfs-kernel.bin
   file, and substitute it below, instead of where I have "LeNgTh"
5. Issue the following command, and upload this file using YMODEM protocol
   load -r -v -b 0x80060000 -m ymodem
6. Issue the following commands, substituting as mentioned above:
   fis create -b 0x80060000 -l LeNgTh vmlinux_2
   load -r -v -b 0x80100000 -m ymodem
7. Using YMODEM, upload openwrt-ath79-generic-*-squashfs-rootfs.bin
8. Issue the "fis free" command, and for the first range in its response,
   use a hexadecimal calculator to subtract the start from the end in order
   to substitute it below, with the leading "0x" to specify it in
   hexadecimal, instead of where I have "LeNgTh"
9. Issue the following commands, substituting as mentioned above:
   fis create -b 0x80100000 -l LeNgTh -e 0 -r 0 rootfs
   reset
10.Wait for the status LED to go solid green

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>
Tested-by: Brian Gonyer <bgonyer@gmail.com>

ynezz commented:
@danielg4 We're now on 4.14 so the PR at the current state is not possible to merge or even start the review. As nobody has touched this PR for quite time, I would recommend you to try add this device to ath79/DTS(ar71xx successor) platform instead, which would probably increase the chances for merge.

danielg4 commented:
For now, I've updated the PR to be for 4.14 only, but still ar71xx, and it works correctly, including the GPIO's. Asking for merge in time for the 19.01 release, since that is the final release with ar71xx, in case I don't get the ath79 edition ready in time

danielg4 commented:
There have been multiple, less-trivial changes to ar71xx code merged in the 2 months since I was told it's in maintenance mode only. With a 19.03 branch just around the corner, I am renewing a request for merge.

danielg4 commented:
Please replace the target/ar71xx tag with the target/ath79 tag. Requesting merge both into master and into the 19.07 branch.

ynezz commented on target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c:
```diff
@@ -581,6 +581,7 @@ static void ath79_mii0_ctrl_set_if(struct ag71xx *ag)
 		mii_if = AR71XX_MII0_CTRL_IF_GMII;
 		break;
 	case PHY_INTERFACE_MODE_RGMII:
+	case PHY_INTERFACE_MODE_RGMII_ID:
```
Please make this changes separate commit.

danielg4 replied:
Such a separate commit to ag71xx could apply to both ar71xx and ath79, allowing for the ar71xx edition of this PR to be re-added on top of the ath79 edition.

ynezz commented on target/linux/ath79/generic/config-default:
```diff
@@ -8,7 +8,7 @@ CONFIG_LEDS_RESET=y
 CONFIG_MICREL_PHY=y
 CONFIG_MTD_REDBOOT_PARTS=y
-CONFIG_MTD_REDBOOT_DIRECTORY_BLOCK=-2
+CONFIG_MTD_REDBOOT_DIRECTORY_BLOCK=-3
```
Please separate commit as well, I would like to know, why do you need this change.

danielg4 replied:
This change is needed because that is the location of the directory block on this device. It's that simple. Ideally, this can even go in target/linux/generic.

ynezz replied:
Ok, then please create separate commit for this ath79 kernel config changes and add there this description.

BTW what is going to happen on devices which rely on the current config `CONFIG_MTD_REDBOOT_DIRECTORY_BLOCK=-2` ? Are you making them unbootable?

danielg4 replied:
> Ok, then please create separate commit for this ath79 kernel config changes and add there this description.
> 
> BTW what is going to happen on devices which rely on the current config `CONFIG_MTD_REDBOOT_DIRECTORY_BLOCK=-2` ? Are you making them unbootable?

If -3 fails, -2 is tried, then -1. Under the current config, it starts with -2, then tries -1, and that's it. The current line in target/linux/generic would have it try -1 only. The ath25 target separately sets this to -3 like I'm doing. The only other place in OpenWrt where it is changed is that patch under ixp4xx where some board needs it to be -5.

ynezz commented on target/linux/ath79/dts/ar7161_adtran_bsap1880.dtsi:
```diff
@@ -0,0 +1,100 @@
+	};
+
+	keys {
+		compatible = "gpio-keys-polled";
```
Please use `gpio-keys`, it's interrupt driven, so preferred.

danielg4 replied:
This part was copied from an existing .dts file. I didn't notice it.

ynezz commented:
> Requesting merge both into master and into the 19.07 branch.

Well, it makes no sense for 19.07, as ath79 is not going to have any images generated in that release.

danielg4 commented:
> > Requesting merge both into master and into the 19.07 branch.
> 
> Well, it makes no sense for 19.07, as ath79 is not going to have any images generated in that release.

So should I add the ar71xx version of this back, with the changes requested here?

ynezz commented:
> > > Requesting merge both into master and into the 19.07 branch.
> > 
> > Well, it makes no sense for 19.07, as ath79 is not going to have any images generated in that release.
> 
> So should I add the ar71xx version of this back, with the changes requested here?

Then we're back at the start, aren't we? The only way to get this merged is ath79 to master branch.

danielg4 commented:
> > > > Requesting merge both into master and into the 19.07 branch.
> > > 
> > > 
> > > Well, it makes no sense for 19.07, as ath79 is not going to have any images generated in that release.
> > 
> > 
> > So should I add the ar71xx version of this back, with the changes requested here?
> 
> Then we're back at the start, aren't we? The only way to get this merged is ath79 to master branch.

I mean in addition to that…

adrianschmutzler commented on target/linux/ath79/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac:
```diff
@@ -13,6 +13,11 @@ PHYNBR=${DEVPATH##*/phy}
+	adtran,bsap1800v2|\
+	adtran,bsap1840)
+		base_mac=$(mtd_get_mac_binary "Board data" 2)
+		echo $(macaddr_add "$base_mac" $(($PHYNBR * 8 + 1))) > /sys${DEVPATH}/macaddress
```
Hi, useless echo. Replace with:
`macaddr_add "$base_mac" $(($PHYNBR * 8 + 1)) > /sys${DEVPATH}/macaddress`

danielg4 replied:
Oops. You're right.

danielg4 replied:
Wait, why is the rest of the file like that too?

adrianschmutzler replied:
I don't know. I have sent a patch for that earlier today:
https://patchwork.ozlabs.org/patch/1120551/

adrianschmutzler replied:
Patch has been accepted: https://github.com/openwrt/openwrt/commit/b16faf9bc725134c54a5976d979e94c9feb75446

danielg4 replied:
I'm still waiting for @ynezz to respond to the other comments before I push an update.

ynezz replied:
Sorry, but I don't see any pending comments from you. ar71xx is simply gone, don't bother with that.

danielg4 committed:
ath79: support "rgmii-id" PHY mode in ag71xx

The ag71xx code did not include a case for the "rgmii-id" PHY mode in the
code. There are devices that need this mode, so I'm adding it.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>

danielg4 committed:
ath79: widen RedBoot directory search and add PHY

The Bluesocket BSAP1880 board has a Vitesse PHY, for which the driver was
not being included, and its RedBoot directory block is earlier in the
flash than the search was allowing. This commit prepares for supporting it.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>

danielg4 commented:
This push also corrects a previously overlooked logic error in the first commit.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_adtran_bsap1800v2.dts:
```diff
@@ -0,0 +1,9 @@
+
+/ {
+	model = "Adtran/Bluesocket BSAP-1800v2";
+	compatible = "adtran,bsap1800v2", "qca,ar7161";
```
Just saw this: If v2 is a revision and not really included in the device's name, the proper way would be:
´´´
	model = "Adtran/Bluesocket BSAP-1800 v2";
	compatible = "adtran,bsap1800-v2", "qca,ar7161";
´´´
... so with space and hyphen.

This would obviously require updates in the other files

adrianschmutzler commented on target/linux/ath79/dts/ar7161_adtran_bsap1800v2.dts:
```diff
@@ -0,0 +1,9 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "ar7161_adtran_bsap1880.dtsi"
```
It's a little misleading that you derive 1840 and 1800 DTSes from a DTSI which has 1880. Consider using 18x0 there.

danielg4 replied:
So how would this board potentially be disambiguated from the BSAP-1800 v1, then?

adrianschmutzler replied:
Don't know (There is no 1800v1 in ath79). Do we have to? What would be the difference?

danielg4 replied:
There was a v1 model which had an older "1800" board, in contrast to this "1880" board.

adrianschmutzler replied:
Hmm, that's ugly. I still think the 1880 is misleading, but I'm short of a better idea. So keep it (also in the Makefile).
However, I would still do the v2 space/hyphen rename mentioned above.

adrianschmutzler commented on target/linux/ath79/dts/ar7161_adtran_bsap1880.dtsi:
```diff
@@ -0,0 +1,100 @@
+
+	keys {
+		compatible = "gpio-keys";
+		poll-interval = <20>;
```
poll-interval should be useless if gpio-keys is used?!

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -84,6 +84,30 @@ define Device/8dev_carambola2
 endef
 TARGET_DEVICES += 8dev_carambola2
 
+define Device/adtran_bsap1880
```
Same here: I would go for 18x0 or 1800, otherwise it's misleading.

danielg4 replied:
This board had the 1880 product number, even if that does not *seem* consistent with the model numbers it was used in. I'm inclined to NAK this request for now.

danielg4 commented:
Still requesting this to be applied to the 19.07 branch after it gets merged…

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -84,6 +84,30 @@ define Device/8dev_carambola2
+
+define Device/adtran_bsap1800-v2
+  $(Device/adtran_bsap1880)
+  DEVICE_TITLE := Adtran/Bluesocket BSAP-1800 v2
```
Please update to use DEVICE_VENDOR, DEVICE_MODEL and DEVICE_VARIANT as introduced in:
https://github.com/openwrt/openwrt/commit/8c3dbaf8b49233e275b2dce4e30dc0a2bf8c86c5

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -84,6 +84,30 @@ define Device/8dev_carambola2
+
+define Device/adtran_bsap1840
+  $(Device/adtran_bsap1880)
+  DEVICE_TITLE := Adtran/Bluesocket BSAP-1840
```
Please update to use DEVICE_VENDOR, DEVICE_MODEL and DEVICE_VARIANT as introduced in:
https://github.com/openwrt/openwrt/commit/8c3dbaf8b49233e275b2dce4e30dc0a2bf8c86c5

adrianschmutzler commented:
I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.

bgonyer commented:
No result was returned.

```root@OpenWrt:/# find /proc/device-tree/ -name "*mac-address*"```
```root@OpenWrt:/# ```

adrianschmutzler commented:
> No result was returned.

Thanks, that matches my expectations based on the DTS file.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

danielg4 commented:
Still requesting merge into the 19.07 branch as well, with 598a1e3 in place of cec3f54.

adrianschmutzler commented:
@danielg4 @bgonyer 
Can you tell me which device bears the MAC address from label/sticker at runtime?
Is the correct MAC address retrieved by this one: `$(mtd_get_mac_binary "Board data" 2)` ?

bgonyer commented:
eth0 matches the MAC address on the label outside the unit.

Yes, the correct MAC address is retrieved by `$(mtd_get_mac_binary "Board data" 2)`
